### ISSUE TITLE
What does refactor-nrepl do for initialization?

### ISSUE THREAD

**Issue by ozjongwon** [clojure-emacs/refactor-nrepl#128]
I have a weired problem and need to understand what refactor-nrepl does just after cider-jack-in                                 

This is my minimal profile (commented out some expressions for trouble shooting)

```
{:user {:plugins [[cider/cider-nrepl "0.9.1"]
                          [refactor-nrepl "1.1.0" :exclusions [org.clojure/tools.nrepl]]
                  #_
                  [lein-ancient "0.6.8"]]
        :dependencies [[org.clojure/tools.nrepl "0.2.11"]
                       #_[acyclic/squiggly-clojure "0.1.4"]]}
}
```

With refactor-nrepl enabled like above definition, the underlying java process consumes 100% < cpu for extra over 30 seconds(compare to refactor-nrepl disabled profile).

I do not conclude anything yet, but when refactor-nrepl is enabled, after heavy cpu consumption, somehow (at least) a top level atom def (which is initialized with top-level function call) just reinitialized/back(?) to the very first initial value(which is a empty map)

Could anyone explain me what it does for the extra 30 seconds and where to look at to understand what it does?

Thanks!

**Comment by ozjongwon** [comment]
Below .emacs config fixes the problem. Is the above behaviour expected?

```
(setq cljr-eagerly-build-asts-on-startup nil
      cljr-auto-clean-ns nil)
```

**Comment by benedekfazekas** [comment]
right after the REPL starts up refactor-nrepl builds ASTs for your namespaces to enable the more advanced features like find usages, extract function, rename symbol to name a few. we use `tools.analyzer` for this especially one of its functions which also `eval`s first level forms while building the AST. This is necessary to really understand your code for these features.

as u rightly guessing you can opt out for eagerly building the ASTs however we will still build them when they are required on tge fly for example if you run a global rename symbol.

This has the undesirable side effect of the reinitialization of the atom you are mentioning. however, initialize an atom like this might not be the best approach ; that said i of course have no idea of the context of ur application so it might be the best available there.

**Comment by expez** [comment]
>  Is the above behaviour expected?

Yeah, unfortunately it is.  You might wonder why this is acceptable at all, and the reason is that most clojure programmers don't keep global state, like your problematic atom, so they're unaffected by this problem.

In general you can avoid global state by adopting a more functional style and/or by using patterns like the [component](https://github.com/stuartsierra/component) pattern to structure your applications.

In this specific instance you can fix your problem by using a `defonce` to define your atom.

**Comment by ozjongwon** [comment]
Sure, I do know what I do, unfortunately refactor-nrepl does not know what it does.

A program that produces incorrect results twice as fast is infinitely slower.
        — John Osterhout

**Comment by benedekfazekas** [comment]
@expez @ozjongwon these two last comments just make me sad.

I do think that clj-refactor brings value and also think that there is an interesting problem here. So shall we cool heads and discuss a bit further?

@ozjongwon does my above explanation make sense?

**Comment by ozjongwon** [comment]
Well, it is not about make sense or not. It is about correctness. I use Clojure as its creator intended. Th

**Comment by benedekfazekas** [comment]
I don't question your (or anybody's) ability to write good clojure code and hopefully this issue is not about that either. I would like to focus at the problem here: so in the terms of the problem: does the explanation and why `defonce` is a fix here make sense?

**Comment by ozjongwon** [comment]
It is not about defonce fix the problem or not. It is about a program's correctness. As I said, the fact is refactor-nrepl does wrong thing to optimize something. Whatever it does a software development tool must not invalidate valid programs silently.

**Comment by magnars** [comment]
While I agree with your sentiment, @ozjongwon, this isn't about optimization. Because of macros, your program cannot be understood fully by tools without evaluating it. So, to use powerful tooling, you must work around that inherent problem in Clojure.

**Comment by benedekfazekas** [comment]
quoting from bronsa (maintainer of tools.analyzer and tools.reader):

> There's also an analyze+eval function that, as the name suggests, evaluates the form after its analysis and stores the resulting value in the :result field of the AST, this function should be used when analyzing multiple forms, as the analysis of a clojure form might require the evaluation of a previous one to make sense.

from [here](https://github.com/clojure/tools.analyzer.jvm)

so there is a really interesting problem here (sorry for repeating myself)

**Comment by ozjongwon** [comment]
@magnars That is interesting. I used to use a very powerful tool with Common Lisp in the past which did more than what cider+nrepl+... do correctly. Frankly I do not know if it is just limitation of JVM or Clojure or tools.

And I do not have any emotion about this problem - just a plain simple correctness, that is all my concern. That's why I simply asked 'is the behaviour expected?' then when I got 'yes' close it.
